server/jobs/storage_backend.go

```diff
@@ -11,12 +16,112 @@ type StorageBackend interface {
 	Read(key string) ([]string, error)
 
 	// Write logs to the storage backend
-	Write(key string, logs []string) (success bool, err error)
+	Write(key string, reader io.Reader) (success bool, err error)
+}
+
+type storageBackend struct {
+	location      stow.Location
+	logger        logging.SimpleLogging
+	containerName string
+}
+
+func (s *storageBackend) Read(key string) ([]string, error) {
+
+	logs := []string{}
+	err := stow.WalkContainers(s.location, s.containerName, 100, func(container stow.Container, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// Skip if not right container
+		if container.Name() != s.containerName {
+			return nil
+		}
+
+		err = stow.Walk(container, key, 100, func(item stow.Item, err error) error {
+			if err != nil {
+				return err
+			}
+
+			// Found the right object
+			if item.Name() == key {
+
+				r, err := item.Open()
+				if err != nil {
+					return err
+				}
+
+				buf := new(strings.Builder)
+				_, err = io.Copy(buf, r)
+				if err != nil {
+					return err
+				}
+
+				logs = strings.Split(buf.String(), "\n")
+				return nil
+			}
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return []string{}, err
+	}
+	return logs, nil
 }
 
-func NewStorageBackend(jobs valid.Jobs) StorageBackend {
-	// No storage backend configured, return Noop for now
-	return &NoopStorageBackend{}
+func (s *storageBackend) Write(key string, reader io.Reader) (success bool, err error) {
+	s.logger.Info("Writing: %s to bucket: %s", key, s.containerName)
+	err = stow.WalkContainers(s.location, s.containerName, 100, func(container stow.Container, err error) error {
+		if err != nil {
+			s.logger.Info(err.Error())
+			return err
+		}
+
+		// Skip if not right container
+		if container.Name() != s.containerName {
+			return nil
+		}
+
+		_, err = container.Put(key, reader, 100, nil)
+		if err != nil {
+			s.logger.Info("error uploading to s3: ", err)
+			return err
+		}
+		s.logger.Info("successfully uploaded logs for job: %s", key)
+
+		return nil
+	})
+
+	if err != nil {
+		return false, err
+	}
+	return true, nil
```

Comment:
So, the reason we return success and error is because how we've configured the job store to use these values. When persistence is not configured, we keep the jobs in memory. 
https://github.com/lyft/atlantis/blob/688a360231828f3a5d1b95884ac79a81392cbd6c/server/jobs/job_store.go#L136-L144
The `NoopStorageBackend` implementation returns false, nil which means we keep the job in memory. Similarly, the `defaultStorageBackend` returns true, nil when we successfully persist the job. We could use custom error `StorageBackendNotSet` and persist the job if we get this error. However, I don't think using errors for conditional branching is a good approach. What do you think?

Reply:
Thanks for the catch, I have a boolean flag to make sure we found the right container. If we don't, we return a container not found error.

Reply:
based on current implementation the only time you get to cleaning up the buffer is if you already successfully uploaded the object in s3. Having that flag makes it look like there is additional scenario where upload was unsuccessful but there we no errors. If we just assume that error is success/failure criteria for backend's persistence, then we call `delete(j.jobs, jobID)` if there are no errors. No need for if conditions.

Reply:
If we only use the error as success/failure criteria, we will be logging the error even if persist fails when it's not configured: 
https://github.com/lyft/atlantis/blob/3b6da9665b35406aa41b20067df78440fd23e01e/server/jobs/project_command_output_handler.go#L152-L155
I have created a custom error for when storage backend is not configured and check for the error before deleting the buffer and returning it. Lmk what you think of this approach or if you have a simpler one in mind!

Reply:
ok, I missed that part. If that's the case we should either move persistence logic into a decorator or revert to dual return interface. Checking for error type is more confusing. I think for simplicity you can revert back to how it was for now.

Reply:
Yeah I agree! We could use a decorator too but since we delete the job in memory only when persist succeeds, we'll have to go to the JobStore again and delete the job from memory which adds complexity imo. I'll revert it back to how it was for now!